src/poll_guard.rs

```diff
@@ -0,0 +1,109 @@
+
+use anyhow::Result;
+use walkdir::{DirEntry, WalkDir};
+
+/// OPTIMIZATION for checking for changes
+///
+/// Provides a function, dir_changed, that is a much faster way to detect if any files in
+/// a repository have changed, vs the naive method of trying to commit the repo. This peeks at
+/// the file timestamp, which is typically cached in memory. The previous way to do it was to
+/// let Git2 make a commit, which triggered a whole lot of I/O and hashing.
+pub struct PollGuard {
+    git_cache: HashMap<PathBuf, Repository>,
+}
+
+impl PollGuard {
+    pub fn new() -> Self {
+        Self {
+            git_cache: Default::default(),
+        }
+    }
+
+    pub fn dir_changed(&mut self, dir: &Path) -> bool {
+        let watermark = match self.get_watermark(dir) {
+            Ok(watermark) => watermark,
+            // True because we want to turn off this optimization
+            Err(_) => return true,
+        };
+
+        fn compare_times(modified: SystemTime, watermark: SystemTime) -> Result<bool> {
+            let duration = modified.duration_since(watermark)?;
+            Ok(duration.as_secs_f32() > 1.0)
+        }
+
+        fn get_file_time(entry: walkdir::Result<DirEntry>) -> Result<SystemTime> {
+            Ok(entry?.metadata()?.modified()?)
+        }
+
+        for entry in WalkDir::new(dir) {
+            let repo = self.git_cache.get(dir).unwrap();
+            if let Ok(ref repo_path) = entry {
+                if repo.is_path_ignored(repo_path.path()).unwrap_or(false) {
+                    continue;
+                }
+            }
+
+            if let Ok(modified) = get_file_time(entry) {
+                if compare_times(modified, watermark).unwrap_or(false) {
+                    dbg!(modified, watermark);
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    /// Find the last known commit timestamp
+    fn get_watermark(&mut self, path: &Path) -> Result<SystemTime> {
+        let repo: &Repository = match self.git_cache.get(path) {
+            Some(found) => found,
+            None => {
+                let new = Repository::open(path)?;
+                self.git_cache.insert(path.to_path_buf(), new);
+                self.git_cache.get(path).unwrap()
+            }
+        };
+
+        fn get_time(commit: &Commit) -> SystemTime {
+            SystemTime::UNIX_EPOCH.add(Duration::from_secs(commit.time().seconds() as u64))
+        }
+
+        fn get_dura_time(head: &Commit, repo: &Repository) -> Result<SystemTime> {
+            let branch_name = format!("dura/{}", head.id());
+            let ret = repo
+                .find_branch(&branch_name, BranchType::Local)?
+                .get()
+                .peel_to_commit()?;
+            Ok(get_time(&ret))
+        }
+
+        // get commit time and fallback to time of HEAD
+        let head = repo.head()?.peel_to_commit()?;
+        get_dura_time(&head, repo).or_else(|_| Ok(get_time(&head)))
```

Comment:
I don't understand this one. This is a fallback chain, I don't want to unwrap, I want to provide an alternative

Reply:
`or_else` and `-unwap_or_else` are quite similar and probably not brilliantly named, but `unwrap_or_else` is precisely for fallback values. The difference is `or_else` takes a closure that returns `Result<T>`, whereas `unwrap_or_else` takes a closure that returns just `T`. 

Because your closure is always returning an `Ok` variant, and is not doing anything that could ever return an `Err` variant, you could use `unwrap_or_else` instead:

```suggestion
        get_dura_time(&head, repo).unwrap_or_else(|_| get_time(&head))
```